File: core/src/main/java/com/serious/service/ServiceInstanceRegistry.java
package com.serious.service;
/*
 * @COPYRIGHT (C) 2016 Andreas Ernst
 *
 * All rights reserved
 */

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cloud.client.ServiceInstance;
import org.springframework.stereotype.Component;

import java.net.URI;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * @author Andreas Ernst
 */
@Component
@Slf4j
public class ServiceInstanceRegistry {
    // local classes

    public static class Delta {
        // instance data

        Map<String,List<ServiceInstance>> deletedInstances = new HashMap<>();
        Map<String,List<ServiceInstance>> addedInstances = new HashMap<>();

        // private

        List<ServiceInstance> getDeletedInstances(String service) {
            return deletedInstances.computeIfAbsent(service, k -> new LinkedList<>());
        }

        List<ServiceInstance> getAddedInstances(String service) {
            return addedInstances.computeIfAbsent(service, k -> new LinkedList<>());
        }

        // public

        public boolean isDeleted(ServiceInstance instance) {
            for (List<ServiceInstance> instances : deletedInstances.values())
                if ( instances.contains(instance))
                    return true;

            return false;
        }

        public void deletedServices(String service, List<ServiceInstance> instances) {
            getDeletedInstances(service).addAll(instances);
        }
        public void deletedService(String service, ServiceInstance instance) {
            getDeletedInstances(service).add(instance);
        }

        public void addedService(String service, ServiceInstance instance) {
            getAddedInstances(service).add(instance);
        }
        public void addedServices(String service, List<ServiceInstance> instances) {
            getAddedInstances(service).addAll(instances);
        }

        public boolean isEmpty() {
            return addedInstances.isEmpty() && deletedInstances.isEmpty();
        }

        public void report() {
            for ( String service : deletedInstances.keySet())
                for (ServiceInstance instance : deletedInstances.get(service))
                    System.out.println("delete " + service + " instance " + instance.toString());

            for ( String service : addedInstances.keySet())
                for (ServiceInstance instance : addedInstances.get(service))
                    System.out.println("added " + service + " instance " + instance.toString());
        }
    }

    // instance data
    @Autowired
    ChannelManager channelManager;
    @Autowired
    ComponentRegistry componentRegistry;
    private Map<String, List<ServiceInstance>> serviceInstances = new ConcurrentHashMap<>();

    // public

    public void startup() {
        // fill initial services

        for (ComponentDescriptor componentDescriptor : ComponentDescriptor.descriptors) {
            List<ServiceInstance> instances = componentRegistry.getInstances(componentDescriptor.getName());

            serviceInstances.put(componentDescriptor.getName(), instances);
        }
    }

    //


    private List<ServiceAddress> extractAddresses(String channels) {
        return  Arrays.stream(channels.split(","))
                .map(channel -> {
                    int lparen = channel.indexOf("(");
                    int rparen = channel.indexOf(")");

                    String protocol = channel.substring(0, lparen);
                    URI uri = URI.create(channel.substring(lparen + 1, rparen));

                    return new ServiceAddress(protocol, uri);
                })
                .collect(Collectors.toList());
    }

    private List<ServiceInstance> getInstances(ComponentDescriptor<?> componentDescriptor) {
        List<ServiceInstance> instances = this.serviceInstances.get(componentDescriptor.getName());

        return instances != null ? instances : new ArrayList<>();
    }
    //

    List<ServiceAddress> getServiceAddresses(ComponentDescriptor<?> componentDescriptor, String... preferredChannels) {
        List<ServiceInstance> instances = getInstances(componentDescriptor);
        Map<ServiceInstance, List<ServiceAddress>> addresses = new HashMap<>();

        // extract addresses

        for ( ServiceInstance instance : instances)
            addresses.put(instance, extractAddresses(instance.getMetadata().get("channels")));

        // figure out a matching channel

        String channelName = null;
        if (!instances.isEmpty()) {
            if (preferredChannels.length == 0) {
                channelName = addresses.get(instances.get(0)).get(0).getChannel();
            }
            else {
                for (ServiceInstance instance : instances) {
                    for ( ServiceAddress address : addresses.get(instance))
                        if ( channelName == null && Arrays.asList(preferredChannels).contains(address.getChannel()) ) {
                            channelName = address.getChannel();
                            break;
                        }

                    if  (channelName != null)
                        break;
                } // for
            }
        }

        if (channelName != null) {
            String finalChannelName = channelName;

            return instances.stream()
                    .filter(serviceInstance ->  addresses.get(serviceInstance).stream().anyMatch(address -> address.getChannel().equals(finalChannelName)))
                    .map(serviceInstance -> new ServiceAddress(finalChannelName, serviceInstance))
                    .collect(Collectors.toList());
        }
        else return null;
    }

    public void report() {
        StringBuilder builder = new StringBuilder();

        builder.append("Service Instances").append("\n");

        for (String service : serviceInstances.keySet()) {
            List<ServiceInstance> instances = serviceInstances.get(service);

            builder.append("Service ").append(service).append("\n");

            for (ServiceInstance instance : instances) {
                builder
                        .append("\t").append(instance.getUri())
                        .append("\n");
            } // for
        } // for

        System.out.println(builder);
    }



    private Delta computeDelta(Map<String, List<ServiceInstance>> oldMap, Map<String, List<ServiceInstance>> newMap) {
        Delta delta = new Delta();

        Set<String> oldKeys = oldMap.keySet();
        for (String service : oldKeys) {
            if (!newMap.containsKey(service))
                delta.deletedServices(service, oldMap.get(service));//System.out.println("deleted service " + service);

            else {
                // check instances

                Map<String, ServiceInstance> oldInstances = new HashMap<>();
                Map<String, ServiceInstance> newInstances = new HashMap<>();

                for ( ServiceInstance serviceInstance : oldMap.get(service))
                    oldInstances.put(serviceInstance.getInstanceId(), serviceInstance);

                for ( ServiceInstance serviceInstance : newMap.get(service))
                    newInstances.put(serviceInstance.getInstanceId(), serviceInstance);

                //oldMap.get(service).stream().peek(serviceInstance -> oldInstances.put(serviceInstance.getInstanceId(), serviceInstance));
                //newMap.get(service).stream().peek(serviceInstance -> newInstances.put(serviceInstance.getInstanceId(), serviceInstance));

                // compare maps

                Set<String> oldInstanceIds = oldInstances.keySet();
                for (String instanceId : oldInstanceIds)
                    if (!newInstances.containsKey(instanceId))
                        delta.deletedService(service, oldInstances.get(instanceId));//System.out.println("deleted service instance " + instanceId);

                for (String instanceId : newInstances.keySet())
                    if (!oldInstanceIds.contains(instanceId))
                        delta.addedService(service, newInstances.get(instanceId));//System.out.println("new service instance " + instanceId);
            }
        }

        for (String service : newMap.keySet())
            if (!oldMap.containsKey(service))
                delta.addedServices(service, newMap.get(service));//System.out.println("new service " + service);

        return delta;
    }

    public void update(Map<String, List<ServiceInstance>> newMap) {
        Delta delta = computeDelta(this.serviceInstances , newMap);

        if (!delta.isEmpty()) {
            delta.report();

            // recheck missing channels

            ChannelInvocationHandler.recheck(channelManager, delta);
        }

        // new map

        this.serviceInstances = newMap;
    }
}

File: core/src/main/java/com/serious/service/ServiceInstanceRegistry.kt
package com.serious.service
/*
* @COPYRIGHT (C) 2023 Andreas Ernst
*
* All rights reserved
*/

import com.serious.service.ChannelInvocationHandler.Companion.updateTopology
import lombok.extern.slf4j.Slf4j
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cloud.client.ServiceInstance
import org.springframework.stereotype.Component
import java.io.Serializable
import java.net.URI
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.stream.Collectors

 /**
 * `ServiceInstanceRegistry` is responsible to cache the current known service instances.
 */
@Component
@Slf4j
class ServiceInstanceRegistry {
    // local classes

    class TopologyUpdate : Serializable {
        // instance data

        private var deletedServices : MutableList<String> = LinkedList()
        private var addedServices   : MutableList<String> = LinkedList()
        private var deletedInstances: MutableMap<String, MutableList<ServiceInstance>> = HashMap()
        private var addedInstances  : MutableMap<String, MutableList<ServiceInstance>> = HashMap()

        // public

        fun involvesService(service : String) : Boolean {
            return deletedInstances.containsKey(service) || addedInstances.containsKey(service)
        }

        fun getDeletedServices(): List<String> {
            return deletedServices
        }
        fun getAddedServices(): List<String> {
            return addedServices
        }

        fun getAddedInstances():Map<String, List<ServiceInstance>> {
            return addedInstances
        }

        fun getDeletedInstances():Map<String, List<ServiceInstance>> {
            return deletedInstances
        }


        fun getDeletedInstances(service: String): MutableList<ServiceInstance> {
            return deletedInstances.computeIfAbsent(service) { _: String? -> LinkedList() }
        }

        fun getAddedInstances(service: String): MutableList<ServiceInstance> {
            return addedInstances.computeIfAbsent(service) { _: String? -> LinkedList() }
        }

        // public
        fun isDeleted(instance: ServiceInstance?): Boolean {
            for (instances in deletedInstances.values)
                if (instances.contains(instance)) return true

            return false
        }

        fun deletedServices(service: String, instances: List<ServiceInstance>) {
            deletedServices.add(service)

            for (serviceInstance in instances)
                deletedService(service, serviceInstance)
        }

        fun deletedService(service: String, instance: ServiceInstance?) {
            log.info(" deleted {} instance {}", service, instance!!.instanceId)

            getDeletedInstances(service).add(instance)
        }

        fun addedService(service: String, instance: ServiceInstance?) {
            log.info(" added {} instance {}", service, instance!!.instanceId)

            getAddedInstances(service).add(instance)
        }

        fun addedServices(service: String, instances: List<ServiceInstance>) {
            addedServices.add(service)

            for (serviceInstance in instances)
                addedService(service, serviceInstance)
        }

        val isEmpty: Boolean
            get() = addedInstances.isEmpty() && deletedInstances.isEmpty()
    }

    // instance data

    @Autowired
    lateinit var channelManager: ChannelManager

    @Autowired
    lateinit var componentRegistry: ComponentRegistry
    private var serviceInstances: MutableMap<String, List<ServiceInstance>> = ConcurrentHashMap()
    private val listener = ArrayList<TopologyListener>();

    // public

     fun addListener(listener: TopologyListener) {
         this.listener.add(listener)
     }

    fun startup() {
        // fill initial services

        val initialInstances: MutableMap<String, List<ServiceInstance>> = ConcurrentHashMap()

        for (componentDescriptor in ComponentDescriptor.descriptors.values)
            initialInstances[componentDescriptor.name] = componentRegistry.getInstances(componentDescriptor.name)

        // first incremental update needed for unresolved injected services

        this.update(initialInstances)
    }

    // private

    private fun extractAddresses(channels: String): List<ChannelAddress> {
        return Arrays.stream(channels.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
            .map { channel: String ->
                val lparen = channel.indexOf("(")
                val rparen = channel.indexOf(")")
                val protocol = channel.substring(0, lparen)
                val uri = URI.create(channel.substring(lparen + 1, rparen))

                ChannelAddress(protocol, uri)
            }
            .collect(Collectors.toList())
    }

     public fun getServices(): List<String> {
         return ArrayList(serviceInstances.keys)
     }

    public fun getInstances(component: String): List<ServiceInstance> {
        val instances = serviceInstances[component]

        return instances ?: emptyList()
    }

    fun getServiceAddress(component: String, preferredChannel: String? = null): ServiceAddress? {
        val instances = getInstances(component)
        val addresses: MutableMap<ServiceInstance, List<ChannelAddress>> = HashMap()

        // extract addresses

        for (instance in instances)
            addresses[instance] = extractAddresses(instance.metadata["channels"]!!)

        // figure out a matching channel

        var channelName: String? = null
        if (!instances.isEmpty()) {
            if (preferredChannel === null) {
                channelName = addresses[instances[0]]!![0].channel
            }
            else {
                for (instance in instances) {
                    for (address in addresses[instance]!!)
                        if (channelName == null && preferredChannel == address.channel) {
                        channelName = address.channel
                        break
                    }
                    if (channelName != null) break
                } // for
            }
        }

        return if (channelName != null) {
            val instances = instances.stream()
                .filter { serviceInstance: ServiceInstance ->
                    addresses[serviceInstance]!!
                        .stream().anyMatch { address: ChannelAddress -> address.channel == channelName }
                }
                .collect(Collectors.toList())

            return ServiceAddress(component, channelName, instances)
        } else null
    }

    fun report() {
        val builder = StringBuilder()

        builder.append("Service Instances").append("\n")

        for (service in serviceInstances.keys) {
            val instances = serviceInstances[service]!!

            builder.append("Service ").append(service).append("\n")

            for (instance in instances) {
                builder
                    .append("\t").append(instance.uri)
                    .append("\n")
            } // for
        } // for

        println(builder)
    }

    private fun topologyUpdate(oldMap: Map<String, List<ServiceInstance>>, newMap: Map<String, List<ServiceInstance>>): TopologyUpdate {
        val topologyUpdate = TopologyUpdate()
        val oldKeys = oldMap.keys
        for (service in oldKeys) {
            if (!newMap.containsKey(service)) topologyUpdate.deletedServices(
                service,
                oldMap[service]!!
            )
            else {
                // check instances

                val oldInstances: MutableMap<String, ServiceInstance> = HashMap()
                val newInstances: MutableMap<String, ServiceInstance> = HashMap()

                for (serviceInstance in oldMap[service]!!)
                    oldInstances[serviceInstance.instanceId] = serviceInstance

                for (serviceInstance in newMap[service]!!)
                    newInstances[serviceInstance.instanceId] = serviceInstance

                // compare maps

                val oldInstanceIds: Set<String> = oldInstances.keys
                for (instanceId in oldInstanceIds)
                    if (!newInstances.containsKey(instanceId))
                        topologyUpdate.deletedService(service, oldInstances[instanceId])

                for (instanceId in newInstances.keys)
                    if (!oldInstanceIds.contains(instanceId))
                        topologyUpdate.addedService(service, newInstances[instanceId])
            }
        }

        for (service in newMap.keys)
            if (!oldMap.containsKey(service))
                topologyUpdate.addedServices(service, newMap[service]!!)

        return topologyUpdate
    }

    fun update(newMap: MutableMap<String, List<ServiceInstance>>) {
        val topologyUpdate = topologyUpdate(serviceInstances, newMap)

        // set new map, so channels can already adapt

        serviceInstances = newMap

        // check for necessary updates

        for ( listener in this.listener)
            listener.update(topologyUpdate) // they are also interested about helath updates?

        if (!topologyUpdate.isEmpty) {
            // inform listener

            updateTopology(this, topologyUpdate)
        }
    }

     companion object {
         val log: Logger = LoggerFactory.getLogger(ServiceInstanceRegistry::class.java)
     }
}
